lib/ts/controllers/s-modal.ts

```diff
@@ -0,0 +1,222 @@
+        static targets = ["modal"];
+
+        private modalTarget!: HTMLElement;
+
+        private _boundClickFn!: any;
+        private _boundKeypressFn!: any;
+
+        private returnElement!: HTMLElement;
+
+        private _boundTabTrap!: any;
+
+        connect () {
+            this._validate();
+        }
+
+        /**
+         * Disconnects all added event listeners on controller disconnect
+         */
+        disconnect() {
+            this._unbindDocumentEvents();
+        };
+
+        /**
+         * Toggles the visibility of the modal
+         */
+        toggle () {
+            this._toggle();
+        }
+
+        /**
+         * Shows the modal
+         */
+        show () {
+            this._toggle(true);
+        }
+
+        /**
+         * Hides the modal
+         */
+        hide () {
+            this._toggle(false);
+        }
+
+        /**
+         * Validates the modal settings and attempts to set necessary internal variables
+         */
+        _validate() {
+            // check for returnElement support
+            var returnElementSelector = this.data.get("return-element");
+            if (returnElementSelector) {
+                this.returnElement = <HTMLElement>document.querySelector(returnElementSelector);
+
+                if (!this.returnElement) {
+                    throw "Unable to find element by return-element selector: " + returnElementSelector;
+                }
+            }
+        }
+
+        /**
+         * Toggles the visibility of the modal element
+         * @param show Optional parameter that force shows/hides the element or toggles it if left undefined
+         */
+        _toggle (show?: boolean | undefined) {
+            var toShow = show;
+
+            // if we're letting the class toggle, we need to figure out if the popover is visible manually
+            if (typeof toShow === "undefined") {
+                toShow = this.modalTarget.getAttribute("aria-hidden") === "true";
+            }
+
+            // show/hide events trigger before toggling the class
+            this.triggerEvent(toShow ? "show" : "hide");
+
+            this.modalTarget.setAttribute("aria-hidden", toShow ? "false" : "true");
+
+            if (toShow) {
+                this._bindDocumentEvents();
+            }
+            else {
+                this._unbindDocumentEvents();
+                this._focusReturnElement();
+            }
+
+            // check for transitionend support
+            var supportsTransitionEnd = (<HTMLElement>this.modalTarget).ontransitionend !== undefined;
+
+            // shown/hidden events trigger after toggling the class
+            if (supportsTransitionEnd) {
+                // wait until after the modal finishes transitioning to fire the event
+                this.modalTarget.addEventListener("transitionend", () => {
+                    //TODO this is firing waaay to soon?
+                    this.triggerEvent(toShow ? "shown" : "hidden");
```

Comment:
I get that the focus trap brings focus into the modal on tab, but I'd still like the ability to automatically focus bring focus into the modal when it appears.

e.g. `data-target="s-modal.initialFocus"`...

Focusing would look like `if (this.initialFocusTargets.length > 0) { this.initialFocusTargets[0].focus(); }`, assuming that the targets array is in DOM order.

Reply:
I'd also like to see `data-s-modal-remove-when-hidden` which removes the element from the DOM after triggering the final event.  A lot of our modals are short-lived, getting generated by the server, doing their job once, and then disappearing forever.

Reply:
I was just talking about this today with another dev (the focus bit). I can add that easily enough. The second item should also be fairly easy to add as well and would save some boilerplate code across the platform that listens on `hidden` and then removes.

Reply:
Added both features! If `data-s-modal-remove-when-hidden="true"` is set, then it'll remove on `hidden` is triggered. If `initialFocus` is set, then that element will be focused on show, else the first focusable element will be focused instead.